### PR TITLE
Add support for Tronxy V10 with TFT_TRONXY_X5SA and MKS_ROBIN_TFT43 screens

### PR DESCRIPTION
**Pull request by smokeypell**
### Description

Add fixes for Tronxy V10 boards using TFT_TRONXY_X5SA and MKS_ROBIN_TFT43 screens.

#### This PR is specifically to add features for the Tronxy V10 boards listed below.
- CXY-V6-191121 - Found in older models of the XY-3 Pro v2 and X5SA series
- CXY-446-V10-220413 - Found in newer models of the X5SA series

### Benefits
- #### HAL\STM32\tft\tft_fsmc.h
- #### HAL\STM32\tft\tft_fsmc.cpp
  - Added support for FMC (Flexible Memory Controller on STM32F446) devices so that the below screens will function.
    - TFT_TRONXY_X5SA
    - MKS_ROBIN_TFT43
    
- #### ini\STM32f4.ini
  - Added the following build environments, based on information from Tronxy, to compile with and without usb flash drive support, respectively.
    - env:TRONXY_CXY_446_V10
    - env:TRONXY_CXY_446_V10_usb_flash_drive
    
- #### buildroot\share\PlatformIO\scripts\tronxy_script.py
  - Creates compiled fmw_tronxy.bin and fmw_tronxy.hex firmware files and places them in a folder labeled "update" in the root of the Marlin project folder for easy access for the user.
  - The Tronxy V10 board requires the firmware to be named in this manner and inside the folder "update" on the SD card for flashing. This saves the user a few steps.
  
- #### Marlin\src\pins\stm32f4\pins_TRONXY_CXY_446_V10.h
  - Renamed pin file for clarity.
  - Properly defined pins and options for the board to work with the screens.
  
### Configurations

[Configurations.zip](https://github.com/MarlinFirmware/Marlin/files/14079318/Configurations.zip)

### Related Issues
I believe this should resolve the following feature request.

### CODE REVIEW

**Comment by thinkyhead** [comment]
Thanks for the submission. This currently breaks the build for Lerdge boards so we can't merge this until that is fixed.

Meanwhile, the pins file contains a lot of logic concerning flags that are presumably only included in the Tronxy example configurations. Such flags could make it difficult to repurpose the board for other machines. We want to make sure that boards are defined as generically as possible. So we want to avoid specifically defining a "Z2" stepper that could be used as an extra extruder, a rotary axis, as an X2 or Y2 stepper, etc. The simplest way to handle this is to define the extra stepper driver as the next available extruder stepper (e.g., E1) and then the stepper can be automatically assigned to X2, Y2, or Z2 according to the user configuration. A similar principle holds for other pins. So the more generic that the pins file can be, the better.

To sum up, we'll need to continue to fix this up until it doesn't break other builds and to the point where the pins are defined as generically as possible for reuse of the board. I look forward to these items being fixed so that we can merge this submission and improve our support for Tronxy machines.

**Comment by smokeypell** [comment]
@thinkyhead 
I will definitely work on the pins file. I was trying to be as inclusive as possible. I see your point on the specific flags.
I will work on the Lerdge board build errors as well. 

Do I need to close this PR and submit a new one when done?

**Comment by smokeypell** [comment]
I need to give credit to @TheThomasD and @jmz52 for me being able to figure this out. I dug through and used code from their repositories to get this sorted.

Next on the list is the pin file.

**Comment by TheThomasD** [comment]
@smokeypell Thank you for the work! Would be great to see the different Tronxy printers being supported in the standard. I don’t have a lot of time available, but if you need some info, let me know!

**Comment by thinkyhead** [comment]
> Do I need to close this PR and submit a new one when done?

No, just keep adding more commits. We'll squash them all together into a single commit when we merge it.

**Comment by smokeypell** [comment]
I like the FMC_OR_FSMC macro. That's slick.

File: Marlin/src/pins/pins.h [hunk 1]
```diff
@@ -1015,6 +1016,8 @@
     #error "BOARD_LINUX_RAMPS is now BOARD_SIMULATED. Please update your configuration."
   #elif MB(BTT_MANTA_M4P_V1_0)
     #error "BOARD_BTT_MANTA_M4P_V1_0 is now BOARD_BTT_MANTA_M4P_V2_1. Please update your configuration."
+  #elif MB(BOARD_TRONXY_V10)
```
Comment (thisiskeithb) [inline Marlin/src/pins/pins.h]: Need to drop the `BOARD_` prefix or the `MB()` macro won’t work.

Comment (smokeypell) [reply Marlin/src/pins/pins.h]: Got it. 
Thank you, sir.

**Comment by thinkyhead** [comment]
> I like the `FMC_OR_FSMC` macro. That's slick.

If there was a more marked difference between FSMC and FMC it would've been a bad idea, but as they're nearly identical it works out pretty well.

**Comment by KJ7LNW** [comment]
Hi @smokeypell, 

I just tried the configs in the top of this PR on my X5SA Pro 400 with a v10 board using the latest Marlin from git.  It built fine and serial seems to work.  M115 shows the the build, etc.

However, this is what my display looks like:

image

Do you know what needs to change in order to fix it?  Let me know if you need more information.

Thanks for your help!

-Eric

**Comment by smokeypell** [comment]
@KJ7LNW Sorry. I'm just now seeing this. Can you get a photo of the board? Some have USB flash drive support, some don't. Some build options need to be set different depending on that option. I'm going to build for my machine again to see if I can replicate.

**Comment by KJ7LNW** [comment]
Thanks for looking into that, here is a pic:

TRONXY X5SA 400PRO PCB

[TRONXY X5SA 400PRO PCB]

**Comment by smokeypell** [comment]
@KJ7LNW One thing I see is that you have a newer board with a GD32F427 chip. Everything I've built up to this point is for the STM32F446ZET6 board. I'm guessing that this may be, at least, part of the issue. I have one of these boards, but haven't had time to try to build Marlin for it yet. I will try to start working on this board and see if I can figure something out.

**Comment by KJ7LNW** [comment]
Thanks, let me know if you need anything to help with that.

**Comment by smokeypell** [comment]
@KJ7LNW I'm not having any issues compiling and loading the firmware on my GD32 board. I am noticing a lack of screen artifacts when using TFT_GENERIC. Hit me up over email and I can send you the firmware to try. I'd like to have you compile what I have configured so I can try to fine tune things a bit. The printer I compiled for was the XY3 Pro V2. I don't have my X5SA set up, plus I'd like to see if you see some of the same quirks I see with other settings. Would be nice to have someone other that myself testing for issues. This would be helpful in adding to and fine tuning for another pull request.

**Comment by KJ7LNW** [comment]
emailed.
